### ISSUE TITLE
[SUGGESTION] Do not rely on Docker for local development and testing

### ISSUE THREAD

**Issue by adam-ah** [orchest/orchest#96]
Hi Team,

Thanks for this great idea, I'm loving it. However, for local development, I don't see a technological reason to rely on Docker. Many of our data scientists do not use docker (for all different reasons), and there is no good argument to force them to use it.

Looking at the source quickly, orchest should be able to run without a container. Furthermore, there are obvious pain-points with docker (e.g., https://github.com/orchest/orchest/issues/56).

Could you please make the docker dependency optional?

Many thanks!

**Comment by yannickperrenet** [comment]
> Thanks for this great idea, I'm loving it.

Thanks! That is great to hear.

> However, for local development, I don't see a technological reason to rely on Docker. Many of our data scientists do not use docker (for all different reasons), and there is no good argument to force them to use it.

Even though we require Docker as a dependency, data scientists should/do not need to interact with Docker at all. It is just there as an abstraction. So I would argue that you do not need to force them to use it, just install it. However, I might very well be missing an obvious perspective and would love to better understand where you are coming from. Could you elaborate, possibly with an example, why you need to interact with Docker directly or why you don't want to install Docker in the first place?

> Could you please make the docker dependency optional?

Docker really is at the core of the application and making it optional is not an easy task, therefore we would first need to better understand why making it optional is beneficial. Hope you can help us out there :)

**Comment by adam-ah** [comment]
> Could you elaborate, possibly with an example, why you need to interact with Docker directly or why you don't want to install Docker in the first place?

I'm happy to review the code further to understand why you need to rely on docker. My understanding is that you simply package your application into a container/image, but above that, no real need to use docker (i.e., you do not need require docker orchestration/compose to run your stack). 

Our team in general do not install and use docker (or any services per se). We package everything into a single repo for our projects. ```pip install``` is allowed and relied on (e.g., numpy, scipy, etc), but any dependency that needs to run as a service is very much frowned upon (why you do need it? how many services do I need to run before I can run the meaningful part of the code? When are they running? Etc.)

Every service-dependency we have (and they are _very_ rare), we can start with a "```./start.py```" or something similar. Anything we need to install as a service (docker, mysql, nginx) is essentially black-listed because they are inherently either stateful or not easily added to repo and moved across systems (e.g., at what point it becomes unreasonable to ask someone else to install yet-another-service just to run your simple scipy-pandas stack? Obviously, it is never acceptable to require the installation of a service).

If you feel docker is giving you essential services you need (NB, I would be surprised to learn this is the case as docker in itself does not give provide any services per se), I'm happy to review and suggest other approaches. Most services we use as a dependency we can start some kind of ```./start.py``` script, even if they provide a docker optional installation (which is fine for production release, but not for local development).

**Comment by ricklamers** [comment]
Hi @adam-ah thanks for this suggestion. Happy to chip in here.

It is indeed true that we use Docker to package Orchest itself, but that is actually not the primary reason we use Docker.

Pipelines are defined as a set of connected pipeline steps (a DAG) where each step will execute in its own container. This allows a single pipeline step to define its own environment - at the level of abstraction that Docker supports. This could even mean different CUDA versions or simply language level differences (e.g. different versions of Tensorflow in the same pipeline - perhaps when using someone's legacy model to benchmark your new approach against).

Furthermore, containerization at this level allows for easier parallelization - it becomes easier to in the future distribute execution of pipelines across multiple nodes. In addition, container level isolation also allows for more granular control of resources, CPUs, RAM or even assignment of specific hardware accelerators (GPUs, TPUs). Another reason is that we can map more flexibly to different execution backends in the future (think Kubernetes, or other runtime environments).

In short, it allows for (container) isolation that enables Orchest pipelines to become much more powerful than if we had chosen for process level isolation only.

**Comment by adam-ah** [comment]
@ricklamers thank you for taking the time to add these clarifications - I missed the part that you use containers to run the steps as well.

Let me address the solutions with our typical requirements (and maybe the answer is that orchest is a great product, but not fur us, and that's totally fine too :) 

>  This allows a single pipeline step to define its own environment

Your demo shows running a notebook and a few python scripts in a pipeline. That is a very typical way to preprocess and process data during development, but notebooks are atypical in production. However, I have not seen data scientists needing separate environments for these kinds of scripts during development, which leads to the following step:

> This could even mean different CUDA versions or simply language level differences (e.g. different versions of Tensorflow in the same pipeline - perhaps when using someone's legacy model to benchmark your new approach against).

This is a very atypical case; we benchmark based on our and others' previously reported results, in line with publication standards. Mixing different versions of the same framework in the same pipeline is something we work hard not to do for many reasons (even ```virtualenv``` is frowned upon). Having the need during _development_ to separately maintain different versions of the same library is not a requirement we've seen before.

>  isolation that enables Orchest pipelines to become much more powerful than if we had chosen for process level isolation only

I can't seem to follow this argument. The OS is really good at isolating processes, and the code we are already running without docker is our own, no need to isolate them further. In other words, we are already running these scripts natively on our machines, but an orchestration on top of them would be great. Unfortunately, [Node-RED](https://nodered.org/) wasn't really designed for this, but kind of works.

In sum, I can see how using docker in production is reasonable (e.g., AWS simply runs it for you as-is, but then you also have to ask yourself why not use SageMaker instead), but during development, it hampers, rather than facilitates the goal. 

But again, please don't take it as criticism, and no worries if your project-goal is different from our requirements.

**Comment by ricklamers** [comment]
Really appreciate your in depth response. I can see how some of the advantages (mixing versions in the same pipeline) might not be useful to everyone, I would even agree that generally it isn’t even best practice to do so. But sometimes one cannot get around it when depending on third party code.

Before diving deeper into specifics, I’m curious what downsides you see using containers for the individual steps. What I haven’t read yet is reasons why you don’t want container encapsulation.

Genuinely curious, everything you’re sharing - even if it’s possibly not generally applicable (requirements naturally vary across teams) - is really interesting and useful to us.

**Comment by adam-ah** [comment]
> I’m curious what downsides you see using containers for the individual steps. What I haven’t read yet is reasons why you don’t want container encapsulation.

That's fair, let me try. To begin with, Docker is designed to solve problems we have not encountered in local development. In other words, just because there is docker, it doesn't mean we _need_ to use docker. 

Our team - and I would guess most data scientist - use MacOS for local development. Even if we had problems docker solved, on MacOS docker creates a tremendous amount of problems that we do not want to deal with (e.g., https://github.com/docksal/docksal/issues/249 https://github.com/docksal/docksal/issues/1431 https://github.com/docksal/docksal/issues/1350). Some of these have a workaround (e.g., sync with Mutagen if the dataset is small enough) some do not.

Time is scarce, and we use it to solve problems that we and our clients have and drop everything that does not need to be solved. 
Adding orchestration to our pipeline _is_ a problem we have, with a poor solution. 
Convincing everyone to install / restart / debug / optimise / clean up old containers and images / update docker is hard, and the result does not solve any problems for us; in fact, it would create problems, therefore we decided to not use abstraction when not needed. 
So far, none of us had any issues with not using docker.

**Comment by ricklamers** [comment]
We decided we wanted to have first class support for macOS early on, so it’s good to hear that your team uses it. We have yet to encounter issues - but truth be told that could be because we haven’t looked hard enough.

The issues you listed are related to Docksal (a project running on Docker). Those issues don’t seem a fundamental problem with Docker but more interactions between Docker and how Docksal has implemented various parts of their web development tooling.

In short, it’s our goal to _avoid_ any direct interaction between the data scientist and Docker. It’s there to function as a transparent layer to enable certain functionality _for those who need it_. For the rest, it should get out of the way. Because we too, want to let you just focus on solving the problems that you and your clients care about. In fact, it’s at the very core of why we started the Orchest project.

If in practice we notice Docker does not help, but gets in the way we will no doubt remove it and look for a better solution. So far, it hasn’t let us down. Would be curious to hear from you data scientists if it does for them - if they’re interested in giving Orchest a shot.

**Comment by yannickperrenet** [comment]
I will close this issue for now. We now better understand the reasons for the suggestion and hopefully we have also been able to get our perspective across. 

Thanks for thinking along! Feel free to re-open the issue anytime.
